File: backend/src/models/style.py
"""Pydantic schemas for Style Pillars - Story 2.1 & 2.2.

Defines models for style pillar selection, intensity sliders,
and intensity submission with soft constraint warnings.
Uses 100% Vietnamese tailoring terminology per NFR11.
"""

from pydantic import BaseModel, Field


class IntensitySlider(BaseModel):
    """Configuration for a style intensity slider (FR2).

    Represents a configurable parameter within a style pillar.
    Backend is authoritative for min/max/default values and golden_points.
    """

    key: str = Field(..., description="Unique identifier for slider")
    label: str = Field(..., description="Display label in Vietnamese")
    description: str | None = Field(None, description="Tooltip/help text in Vietnamese")
    min_value: float = Field(0.0, description="Minimum slider value")
    max_value: float = Field(100.0, description="Maximum slider value")
    default_value: float = Field(50.0, description="Default slider position")
    step: float = Field(1.0, description="Slider increment step")
    unit: str | None = Field(None, description="Unit label (e.g., '%', 'cm')")
    golden_points: list[float] = Field(
        default_factory=list,
        description="Artisan golden ratio positions for Haptic Golden Points UI (Story 2.2)",
    )

    model_config = {"from_attributes": True}


class StylePillarResponse(BaseModel):
    """Response schema for a Style Pillar (FR1).

    Represents a design style (e.g., Truyền thống, Tối giản).
    Contains associated intensity sliders for fine-tuning.
    """

    id: str = Field(..., description="Unique style pillar identifier")
    name: str = Field(..., description="Display name in Vietnamese")
    description: str = Field(..., description="Style description in Vietnamese")
    image_url: str | None = Field(None, description="Preview image URL")
    sliders: list[IntensitySlider] = Field(
        default_factory=list,
        description="Intensity sliders for this style"
    )
    is_default: bool = Field(False, description="Whether this is the default style")

    model_config = {"from_attributes": True}


class StylePillarListResponse(BaseModel):
    """Response schema for list of style pillars."""

    pillars: list[StylePillarResponse] = Field(
        default_factory=list,
        description="Available style pillars"
    )
    total: int = Field(0, description="Total number of pillars")

    model_config = {"from_attributes": True}


class StyleSelectionRequest(BaseModel):
    """Request schema for selecting a style pillar."""

    pillar_id: str = Field(..., description="ID of selected style pillar")


class IntensityUpdateRequest(BaseModel):
    """Request schema for updating intensity values."""

    slider_key: str = Field(..., description="Key of the slider to update")
    value: float = Field(..., description="New slider value")


# ===== Story 2.2: Intensity Submission Models =====


class IntensityValueItem(BaseModel):
    """A single slider key-value pair for intensity submission (Story 2.2)."""

    key: str = Field(..., description="Slider key identifier")
    value: float = Field(..., description="Current slider value")


class IntensitySubmitRequest(BaseModel):
    """Request schema for submitting intensity values to backend (Story 2.2).

    Implements sequence-based validation for race condition protection.
    """

    pillar_id: str = Field(..., description="ID of the selected style pillar")
    intensities: list[IntensityValueItem] = Field(
        default_factory=list,
        description="List of slider key-value pairs to submit",
    )
    sequence_id: int = Field(
        ...,
        ge=0,
        description="Monotonically increasing sequence ID for concurrency control",
    )


class IntensityWarning(BaseModel):
    """A soft constraint warning returned by the backend (Story 2.2).

    Soft warnings allow the submission to succeed but alert the user
    about potential issues using Vietnamese tailoring terminology.
    """

    slider_key: str = Field(..., description="The slider key that triggered the warning")
    message: str = Field(..., description="Vietnamese warning message for the user")
    severity: str = Field(
        "soft",
        description="Warning severity: 'soft' allows submission, 'hard' would block (future)",
    )


class IntensitySubmitResponse(BaseModel):
    """Response schema for intensity submission (Story 2.2)."""

    success: bool = Field(..., description="Whether the submission was accepted")
    sequence_id: int = Field(..., description="Echo of the submitted sequence_id")
    warnings: list[IntensityWarning] = Field(
        default_factory=list,
        description="Soft constraint warnings (submission still accepted)",
    )
    error: str | None = Field(
        None,
        description="Error message if success=False (hard constraint violation)",
    )

File: frontend/src/types/style.ts
/**
 * Style Pillar Types - TypeScript types and Zod schemas
 * Story 2.1 & 2.2: Lựa chọn & Tinh chỉnh Trụ cột Phong cách
 *
 * Uses snake_case for API field names to match Backend SSOT.
 */

import { z } from "zod";
import type { FabricResponse } from "@/types/fabric";
import type { MasterGeometrySnapshot } from "@/types/inference";
import type { MasterGeometry, MorphDelta, ConstraintViolation } from "@/types/geometry";

// ===== Intensity Slider Types =====

/**
 * Configuration for a style intensity slider (FR2).
 * Backend provides min/max/default values and golden_points (SSOT).
 */
export interface IntensitySlider {
  key: string;
  label: string;
  description: string | null;
  min_value: number;
  max_value: number;
  default_value: number;
  step: number;
  unit: string | null;
  /** Artisan golden ratio positions for Haptic Golden Points UI (Story 2.2) */
  golden_points: number[];
}

/**
 * Schema for slider value validation
 */
export const sliderValueSchema = z.object({
  key: z.string(),
  value: z.number(),
});

export type SliderValue = z.infer<typeof sliderValueSchema>;

// ===== Style Pillar Types =====

/**
 * Style Pillar response from API (FR1).
 * Represents a design style with its intensity sliders.
 */
export interface StylePillarResponse {
  id: string;
  name: string;
  description: string;
  image_url: string | null;
  sliders: IntensitySlider[];
  is_default: boolean;
}

/**
 * List of style pillars response
 */
export interface StylePillarListResponse {
  pillars: StylePillarResponse[];
  total: number;
}

// ===== Story 2.2: Intensity Submission Types =====

/**
 * A single slider key-value pair for backend submission (Story 2.2).
 */
export interface IntensityValueItem {
  key: string;
  value: number;
}

/**
 * Request payload for submitting intensity values to backend (Story 2.2).
 */
export interface IntensitySubmitRequest {
  pillar_id: string;
  intensities: IntensityValueItem[];
  sequence_id: number;
}

/**
 * Zod schema for validating intensity submission response from backend.
 */
export const intensityWarningSchema = z.object({
  slider_key: z.string(),
  message: z.string(),
  severity: z.string(),
});

export const intensitySubmitResponseSchema = z.object({
  success: z.boolean(),
  sequence_id: z.number(),
  warnings: z.array(intensityWarningSchema),
  error: z.string().nullish(),
});

export type IntensityWarning = z.infer<typeof intensityWarningSchema>;
export type IntensitySubmitResponse = z.infer<typeof intensitySubmitResponseSchema>;

// ===== Design Session State Types =====

/**
 * Current intensity values for sliders.
 * Key is slider key, value is current position.
 */
export interface IntensityValues {
  [key: string]: number;
}

/**
 * Design session state for Zustand store.
 */
export interface DesignSessionState {
  // Selected style pillar
  selected_pillar: StylePillarResponse | null;

  // Current intensity values for sliders
  intensity_values: IntensityValues;

  // Loading states
  is_loading_pillars: boolean;
  is_pillar_selected: boolean;

  // Story 2.2: Submission state
  is_submitting: boolean;
  last_submitted_sequence: number;
  submission_warnings: IntensityWarning[];

  // Story 2.3: Fabric recommendation state
  fabric_recommendations: FabricResponse[];
  is_loading_fabrics: boolean;

  // Story 2.4: Master Geometry translation state
  master_geometry: MasterGeometrySnapshot | null;
  // Story 3.1: Adaptive Canvas visual pattern state
  current_pattern: MasterGeometry | null;
  is_translating: boolean;
  translate_error: string | null;

  // Story 3.2: Current morph delta from morphing engine
  current_morph_delta: MorphDelta | null;

  // Story 3.3: Comparison Overlay state
  is_comparison_mode: boolean;

  // Story 3.4: Lock Design state
  is_design_locked: boolean;
  is_locking: boolean;
  lock_error: string | null;
  locked_design_id: string | null;
  locked_geometry_hash: string | null;

  // Story 4.1b: Guardrail state
  guardrail_status: "passed" | "warning" | "rejected" | null;
  guardrail_warnings: ConstraintViolation[];
  guardrail_violations: ConstraintViolation[];
  last_valid_sequence_id: string | null;
  last_valid_intensity_values: IntensityValues | null;

  // Error state
  error: string | null;
}

/**
 * Design session actions for Zustand store.
 */
export interface DesignSessionActions {
  // Select a style pillar and initialize sliders
  selectPillar: (pillar: StylePillarResponse) => void;

  // Update a single slider value
  updateIntensity: (key: string, value: number) => void;

  // Reset all sliders to default values
  resetToDefaults: () => void;

  // Set loading state
  setLoading: (loading: boolean) => void;

  // Set error state
  setError: (error: string | null) => void;

  // Clear session
  clearSession: () => void;

  // Story 2.2: Submission state management
  setSubmitting: (loading: boolean) => void;
  setSubmissionResult: (sequenceId: number, warnings: IntensityWarning[]) => void;

  // Story 2.3: Fabric recommendation actions
  setFabricRecommendations: (fabrics: FabricResponse[]) => void;
  setLoadingFabrics: (loading: boolean) => void;
  clearFabricRecommendations: () => void;

  // Story 2.4: Master Geometry translation actions
  setMasterGeometry: (snapshot: MasterGeometrySnapshot | null) => void;
  // Story 3.1: Set current visual pattern
  setCurrentPattern: (pattern: MasterGeometry | null) => void;
  setTranslating: (loading: boolean) => void;
  setTranslateError: (error: string | null) => void;
  clearMasterGeometry: () => void;

  // Story 3.2: Set current morph delta
  setCurrentMorphDelta: (delta: MorphDelta | null) => void;

  // Story 3.3: Toggle comparison overlay
  toggleComparisonMode: () => void;

  // Story 3.4: Lock Design actions
  setLocking: (loading: boolean) => void;
  setLockResult: (designId: string, geometryHash: string) => void;
  setLockError: (error: string | null) => void;

  // Story 4.1b: Guardrail actions
  setGuardrailResult: (result: { status: "passed" | "warning" | "rejected"; violations: ConstraintViolation[]; warnings: ConstraintViolation[]; last_valid_sequence_id: string | null }) => void;
  snapBackToSafe: () => void;
  clearGuardrailState: () => void;
}

/**
 * Complete design store type
 */
export type DesignStore = DesignSessionState & DesignSessionActions;
